### PR TITLE
Fixes issue #6061 Add using writes the wrong reference down for a

### PR DESCRIPTION
**Pull request by mkrueger**
framework assembly

### CODE REVIEW

**Comment by mkrueger** [comment]
I'm not sure in which use case this service is used - @Therzok can you provide one ?

The VS.NET implementation loads reflection assemblies and checks if the type exists - even if the name is null. I would really like to avoid that  - looks like a double check. Roslyn has some way of discovering the type -> assembly name therefore I think this check assembly name -> type is not needed.

However if it is I would rather use cecil to look up the name - loading assemblies that way pollutes the app domain and loading may fail due to dependency/framework compatibility issues.

Any comments ?

**Comment by Therzok** [comment]
> I'm not sure in which use case this service is used - @Therzok can you provide one ?

There's 1 scenario that's fixed by this patch, and that's searching reference assemblies without having them referenced in any project. Saw it used in the codeaction, assumed that it was the fix. We currently have that toggled off, sorry for misdiagnosing.

To fix the bug report scenario, we need to normalize the references at MonoDevelop workspace level and change the path when we encounter a GAC assembly:
http://source.roslyn.io/#Microsoft.CodeAnalysis.Workspaces.MSBuild/MSBuild/ProjectFile/ProjectFile.cs,3a3a092b693fd0e7

My test project to repro the issue:
[testusing.zip](https://github.com/mono/monodevelop/files/2458581/testusing.zip)

Open `testconsaddusing` and trigger quickfix on `MessageQueueInstaller`. If that adds a relative reference to System.Messaging, the test should fail.

I've pushed a minor fixup to this branch, so the (currently) unused scenario is working

File: main/src/core/MonoDevelop.Ide/MonoDevelop.Ide.TypeSystem/MonoDevelopFrameworkAssemblyPathResolver.cs [hunk 1]
```diff
@@ -0,0 +1,117 @@
...
+
+			public string ResolveAssemblyPath (ProjectId projectId, string assemblyName, string fullyQualifiedName = null)
+			{
+				if (workspace == null)
+					return null;
+
+				if (!(workspace.GetMonoProject (projectId) is DotNetProject monoProject))
+					return null;
+
+				string assemblyFile = Runtime.SystemAssemblyService.DefaultAssemblyContext.GetAssemblyLocation (assemblyName, monoProject.TargetFramework);
```
Comment (slluis) [inline main/src/core/MonoDevelop.Ide/MonoDevelop.Ide.TypeSystem/MonoDevelopFrameworkAssemblyPathResolver.cs]: Shouldn't this just use the assembly context from the project? That is, `monoProject.AssemblyContext`

**Comment by slluis** [comment]
Can we have unit tests for this based on Marius' example?

**Comment by mkrueger** [comment]
We can test the service however I don't know how this example should help - it's just using a quick fix. Better is to test such services directly there is no need for throwing quick fixes in they may change or get removed or something.

Roslyn is nothing we can control.

**Comment by slluis** [comment]
@Therzok can you please verify?

**Comment by Therzok** [comment]
Issue is still not fixed, seeing relative path to `/Library` being written.

It seems like roslyn is not considering the assemblies as being reference assemblies, but like simple assembly references.

Trying to figure out what is causing them not to be seen as framework reference assemblies by Roslyn.

**Comment by Therzok** [comment]
Discussed with roslyn folks: https://gitter.im/dotnet/roslyn?at=5bc4c2f2e65a634336e25bdf

Adding some code in workspace layer to handle this.

**Comment by Therzok** [comment]
Currently working on adding tests. Two things don't work yet in this version:
* net461 project referencing System.Messaging, on adding reference to net471, it doesn't correct it to a package assembly
* gdk-sharp is a special case to handle. not sure how.

**Comment by Therzok** [comment]
Added unit tests.

The one problem remaining is `glib-sharp` handling, as in, things which are in gac but not in a package.

I don't know how to solve that problem, especially without breaking AssemblyContext API or making it incredibly slow (IO + lots of allocations even on fast path).

**Comment by slluis** [comment]
@Therzok is the glib-sharp issue a big problem? In which cases things won't work?

**Comment by Therzok** [comment]
It writes a relative path to `/Library/Frameworks...` for assemblies which are not included in a SystemPackage.

I'm not sure it's worth pursuing a fix for that, I've only seen glib-sharp, gtk# and co in there.

**Comment by slluis** [comment]
Ok then.

**Comment by slluis** [comment]
Let's resolve the conflicts and merge.

**Comment by Therzok** [comment]
@monojenkins backport release-7.7

**Pull request merged**